pkg/scte35/stream.go

```diff
@@ -0,0 +1,354 @@
+// Copyright 2022 Adrian of Doom
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or   implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+
+package scte35
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+)
+
+const (
+    // pktSz is the size of an MPEG-TS packet in bytes.
+    pktSz = 188
+
+    // bufSz is the size of a read when parsing files.
+    bufSz = 13000 * pktSz
+)
+
+type PacketData struct {
+	PacketNumber int     `json:",omitempty"`
+	Pid          uint16  `json:",omitempty"`
+	Program      uint16  `json:",omitempty"`
+	Pcr          float64 `json:",omitempty"`
+	Pts          float64 `json:",omitempty"`
+}
+
+//Stream for parsing MPEGTS for SCTE-35
+type Stream struct {
+	pktNum   int // packet count.
+	programs []uint16
+	pid2Prgm map[uint16]uint16 //lookup table for pid to program
+	prgm2pcr map[uint16]uint64 //lookup table for program to pcr
+	prgm2pts map[uint16]uint64 //lookup table for program to pts
+	partial  map[uint16][]byte // partial manages tables spread across multiple packets by pid
+	last     map[uint16][]byte // last compares current packet payload to last packet payload by pid
+	Pids
+}
+
+func (stream *Stream) mkMaps() {
+	stream.pid2Prgm = make(map[uint16]uint16)
+	stream.last = make(map[uint16][]byte)
+	stream.partial = make(map[uint16][]byte)
+	stream.prgm2pcr = make(map[uint16]uint64)
+	stream.prgm2pts = make(map[uint16]uint64)
+}
+
+// Decode fname (a file name) for SCTE-35
+func (stream *Stream) Decode(fname string) {
+	stream.mkMaps()
+	stream.pktNum = 0
+	file, err := os.Open(fname)
+	chk(err)
+	defer file.Close()
+	buffer := make([]byte, bufSz)
+	for {
+		bytesread, err := file.Read(buffer)
+		if err != nil {
+			break
+		}
+		for i := 1; i <= (bytesread / pktSz); i++ {
+			end := i * pktSz
+			start := end - pktSz
+			p := buffer[start:end]
+			pkt := &p
+			stream.pktNum++
+			stream.parse(*pkt)
+		}
+	}
+}
+
+func (stream *Stream) mkPcr(prgm uint16) float64 {
```

Comment:
Another case where you can use a longer, more descriptive method name with an all-caps acronym (e.g. `makePCR`)